admin/cluster_setup.py

```diff
@@ -0,0 +1,477 @@
+            for node in cluster.all_nodes:
+                results.append(capture_upstart(reactor,
+                                               node.address,
+                                               remote_logs_file)
+                               )
+        gather_deferreds(results)
+
+        config = _build_config(cluster, options['template'],
+                               options['apps-per-node'])
+        result = yield _configure(reactor, cluster, config)
+
+    except Exception:
+        result = 1
+        raise
+    finally:
+        if options['no-keep']:
+            runner.stop_cluster(reactor)
+        else:
+            if cluster is None:
+                print("Didn't finish creating the cluster.")
+            else:
+                print("The following variables describe the cluster:")
+                environment_variables = get_trial_environment(cluster)
+                for environment_variable in environment_variables:
+                    print("export {name}={value};".format(
+                        name=environment_variable,
+                        value=shell_quote(
+                            environment_variables[environment_variable]),
+                    ))
+                print("Be sure to preserve the required files.")
+
+    raise SystemExit(result)
+
+
+def _build_config(cluster, application_template,  per_node):
+    application_root = {}
+    applications = {}
+    application_root["version"] = 1
+    application_root["applications"] = applications
+    for node in cluster.agent_nodes:
+        for i in range(per_node):
+            name = "app_%s_%d" % (node.private_address, i)
+            applications[name] = deepcopy(application_template)
+
+    deployment_root = {}
+    nodes = {}
+    deployment_root["nodes"] = nodes
+    deployment_root["version"] = 1
+    for node in cluster.agent_nodes:
+        nodes[node.private_address] = []
+        for i in range(per_node):
+            name = "app_%s_%d" % (node.private_address, i)
+            nodes[node.private_address].append(name)
+
+    return {"applications": application_root,
+            "deployment": deployment_root}
+
+
+def _configure(reactor, cluster, configuration):
+    base_url = b"https://{}:{}/v1".format(
+        cluster.control_node.address, REST_API_PORT
+    )
+    certificates_path = cluster.certificates_path
+    cluster_cert = certificates_path.child(b"cluster.crt")
+    user_cert = certificates_path.child(b"user.crt")
+    user_key = certificates_path.child(b"user.key")
+    body = dumps(configuration)
+    treq_client = treq_with_authentication(
+        reactor, cluster_cert, user_cert, user_key)
+
+    def got_all_nodes():
+        d = treq_client.get(
+            base_url + b"/state/nodes",
+            persistent=False
+        )
+        d.addCallback(check_and_decode_json, OK)
+        d.addCallbacks(
+            lambda nodes: len(nodes) >= len(cluster.agent_nodes),
+            lambda _: False
```

Comment:
Because this drops the Failure passed to the errBack, it should log the error.  You can use `twisted.python.log.err` as the error callback (without a lambda), since it accepts a Failure, and returns `None` (which is equivalent to `False` in `loop_until`.